### ISSUE TITLE
Extend Chef.deprecated API so that it is useful for external users to add their own deprecations

### ISSUE THREAD

**Issue by Xorima** [chef/chef#9015]
### Core Chef Resource Checklist

Before suggesting a resource for inclusion please make sure your suggestion meets these criteria for resources built into Chef:
 - [ ] Automates an operating system component that ships by default on systems such as authentication, raid, disk partitions, firewalls, containers, or virtualization systems.
 - [ ] Does not attempt automate 3rd party applications such as database, web, or application servers, which are best suited for cookbooks due to their fast moving nature.

### Describe the resource:

I would like a resource I can include in my recipes to mark them as deprecated, and so that all deprecations appear at the end of the run in a known part of the run.
Currently a lot of people use `Chef::Log.warn` but this is logged in the middle of the run and may be missed, having a deprecation resource would enable cookbook developers to mark parts of a cookbook as deprecated and have expected behaviour as to where they will display to users

### Why should this be included out of the box?:
So I can continue to upgrade cookbooks to Custom Resources ;)

### What operating systems would it run on?
All

### Can We Help You Implement This?:
Yes

**Comment by lamont-granquist** [comment]
deprecations as a resource make no sense at all.

users could use the `Chef.deprecated` API just like we do, the only problem is that there's no useful API for dynamically adding classed to Chef::Deprecated in ways that the number doesn't collide.  we also tightly couple that API to our own website for deprecations which users won't be able to submit issues to.

**Comment by tas50** [comment]
Thanks for submitting this issue. After review by the team, we've decided that this isn't something that we're going to work on ourselves. We receive dozens of new issues a month, which is more than our team can possibly work on. This means we must prioritize work and be honest about what we're able to get done as a team.  That doesn't mean that we think this is a bad idea or isn't a necessary change. It's just not something we'll likely get to given competing priorities. I've applied our `Design Proposal: Accepted. PRs Welcome` label to this issue and closed it out. If you or someone else would like to work on this issue we're open to pull requests. If you'd like to chat further about what that design might look reach out to us in our [#chef-infra-dev](https://chefcommunity.slack.com/archives/C55FCUGBS) slack channel.
